ion/binarywriter.go

```diff
@@ -172,21 +172,24 @@ func (w *binaryWriter) WriteFloat(val float64) error {
 func (w *binaryWriter) WriteDecimal(val *Decimal) error {
 	coef, exp := val.CoEx()
 
-	vlen := uint64(0)
-	if exp != 0 {
-		vlen += varIntLen(int64(exp))
-	}
-	if coef.Sign() != 0 {
-		vlen += bigIntLen(coef)
+	// If the value is 0. (aka 0d0) then L is zero, there are no length or
+	// representation fields, and the entire value is encoded as the single byte 0x50.
+	if coef.Sign() == 0 && int64(exp) == 0 {
+		buf := make([]byte, 0, 0)
+		buf = appendTag(buf, 0x50, 0)
+
+		return w.writeValue("Writer.WriteDecimal", buf)
```

Comment:
You can just use `w.writeTag(0x50, 0)` here. :) The more complicated path of allocating a slice and using appendXxx to fill it in potentially saves some memory allocations if you can calculate ahead of time how big the buf needs to be.

(Which may or may not actually be any faster, someone should probably profile that...)